File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getInorder(self, root, inorder):
        if root:
            self.getInorder(root.left, inorder)
            inorder.append(root.val)
            self.getInorder(root.right, inorder)
            
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        # BruteForce
        inorder = []
        self.getInorder(root, inorder)
        
        for i in range(len(inorder) - 1):
            if inorder[i] >= inorder[i + 1]:
                return False
        return True
        
        

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return bruteForce(root);
    }
private:   
    bool bruteForce(TreeNode* root) {
        vector<int> inorder;
        getInorder(root, inorder);
        
        for(int i = 0; i < inorder.size() - 1; ++i) {
            if(inorder[i] >= inorder[i + 1])
                return false;
        }
        return true;
    }
    
    void getInorder(TreeNode* root, vector<int>& inorder) {
        if(root) {
            getInorder(root -> left, inorder);
            inorder.push_back(root -> val);
            getInorder(root -> right, inorder);
        }
    }
    
    
    
    
    
    bool solve(TreeNode* root) {
        if(root == nullptr)
            return true;
         
        if((root -> left != nullptr && root -> val <= root -> left -> val) || (root -> right != nullptr && root -> val >= root -> right -> val))
            return false;
        
        return solve(root -> left) && solve(root -> right);
    }
};